qiime2/core/type/signature.py

```diff
@@ -109,6 +109,14 @@ def __init__(self, callable, inputs, parameters, outputs,
             Output name to description string.
 
         """
+        # update type of outputs if needed
+        if type(outputs) is list:
+            outputs = dict(outputs)
+        elif type(outputs) is set:
+            raise ValueError("Plugin registration for %r cannot use a set()"
+                             " to define the outputs, as the order is random."
+                             % callable.__name__)
```

Comment:
This was surprisingly easy to have happen without noticing:
```python
plugin.register_function(
...
    outputs={
        ('foo', FooType),
        ('bar', BarType)
    },
...
)
```

does not look particularly wrong, but would result in mis-ordered outputs which randomly shuffled themselves.